buildtest/defaults.py

```diff
@@ -31,11 +35,16 @@
 
 # dictionary used for storing status of builds
 BUILDTEST_BUILD_HISTORY = {}
-BUILDTEST_CONFIG_FILE = os.path.join(BUILDTEST_ROOT, "settings.yml")
-BUILDTEST_CONFIG_BACKUP_FILE = os.path.join(BUILDTEST_ROOT, "settings.yml.bak")
+BUILDTEST_CONFIG_FILE = os.path.join(BUILDTEST_ROOT, "settings.json")
+BUILDTEST_CONFIG_BACKUP_FILE = os.path.join(BUILDTEST_ROOT, "settings.json.bak")
 
 # TESTCONFIG_ROOT is the root directory where test configurations are found
 # configs can be specified as full paths or relative to this path
 TESTCONFIG_ROOT = os.path.join(BUILDTEST_ROOT, "site")
-DEFAULT_CONFIG_FILE = os.path.join(root, "settings.yml")
+DEFAULT_CONFIG_FILE = os.path.join(root, "settings", "settings.json")
+DEFAULT_CONFIG_SCHEMA = os.path.join(root, "settings", "settings.schema.json")
 EDITOR_LIST = ["vim", "emacs", "nano"]
+supported_launcher_dict = {
+    "local": ["local", "mpirun", "mpiexec"],
+    "slurm": ["mpirun", "mpiexec", "srun"],
```

Comment:
@shahzebsiddiqui these configs aside, could you please provide me examples of recipes that would be run in the context of some scheduler? For example, I should be able to take the same "script" schema recipe and then hand it to one of two executors - local or slurm - and my choice should not impact the test config file. What does that look like? E.g.,:

```
buildtest build -c my-script.yml -e slurm 
```
and the default would be local
```
buildtest build -c my-script.yml
```
What is confusing to me is that if a user wants some particular executable (e.g., mpirun) wouldn't they just write this into their script (or other type of test config)? Why does this need to be represented on the level of buildtest? For something like srun this makes sense because we need to launch a job (and then run something from it) but for the rest it isn't clear that this is the right level of representation.

Reply:
It might be helpful if you can point me to a configuration file for each of the examples:

 - a config that defaults to just local
 - a config that expects slurm
 - a config that defaults to using mpi (something)

So I can better understand this modeling. Thank you!

Reply:
that dictionary was intended to check supported launcher type ``launcher: srun`` in configuration with ``scheduler: slurm`` but depending if you change the json format this section won't be needed.

In regards to the test-configuration i foresee ability for test to be run in one or more queues if possible. 

Let's say the test https://github.com/HPC-buildtest/buildtest-framework/blob/devel/examples/script/zlib.yml could have a key to specify one or more queues to run the test. Let's assume a default queue `local` is provided by buildtest we could have test written something like 

``` 
version: 0.0.1

zlib_compression:
  type: script
  queues: ["local"]
  run: |
    dd if=/dev/zero of=1mb.txt bs=1M count=1
    tar cvzf 1mb.tar.gz 1mb.txt
    ls -lh 1mb.tar.gz
    gzip -l 1mb.tar.gz
```

Most general scripts that are shell script for sanity check on system won't need to be run locally. I think some useful examples will come as i start testing,

For instance in stampede2 we have the following queues 

```
login4.stampede2(1001)$ sinfo -o '%R' -h
development
normal
large
long
flat-quadrant
skx-normal
skx-large
skx-dev
```

I may want to run let's say stream test https://github.com/HPC-buildtest/buildtest-framework/blob/devel/examples/benchmarks/stream.yml on one or more queue 
something like. Note the names ``normal`` and ``sky-normal`` would be queue names in buildtest configuration.

```
version: 0.0.1

stream_uniprocess_c:
  type: script
  queues: ["normal", "sky-normal"]
  description: "STREAM Microbenchmark C test on uniprocessor"
  run: |
    wget https://www.cs.virginia.edu/stream/FTP/Code/stream.c
    gcc -O2 stream.c -o stream
    ./stream
```

I think some sites may define a queue for ``gpu`` workload, it makes sense to have GPU centric tests be assigned to those queues or buildtest provides a queue that is configured to setup a GPU environment.

I may want a queue configuration for GPU workload to define let's say a gpu_volta queue to select volta gpus with 5G and automatically do ``module load CUDA/10.0``.  Note that queue is "gpu_volta" but actual partition is called **gpu**. It may allow user to define multiple queues for same partition. 
 
```
 "slurm": {
          "gpu_volta": {
              "description": "gpu queue for volta",
              "launcher": "srun",
              "partition": "gpu",
              "options": [
                "#SBATCH -n 1",  
                "#SBATCH --gres=gpu:volta:1",
                "#SBATCH --mem-per-gpu 5G"
                ]
             "modules": 
                purge: True
                names: ["CUDA/10.0"]
          }
```

Then if i have some gpu test i could just reference the queue to prepoluate the information, and allow some configuration to be overridden in test-configuration.

Let's say a test for saxpy as follows:

```
version: 0.0.1
saxpy:
  type: script
  queues: ["gpu_volta"]
  run: |
  nvcc saxpy.cu -o saxpy
  ./saxpy
```

I can foresee the test have content 

```
#!/bin/bash
#SBATCH -n 1
#SBATCH -p gpu
#SBATCH --gres=gpu:volta:1
#SBATCH --mem-per-gpu 5G
module load CUDA/10.0
nvcc saxpy.cy -o saxpy
./saxpy
```

These are just some ideas and it may take some time to flesh out all the details. I can see myself have to work with site-admins to run benchmark during maintenance period, so a suite of benchmark that are configured to use a high priority queue such as ``admin`` could be configured on the site and all tests are configured with that tag.

Reply:
okay, so you validated one of my concerns - the build recipes should be robust to run _across environments_ without requiring a hard coded executor. So this example:

```
version: 0.0.1

zlib_compression:
  type: script
  queues: ["gpu_queue"]
  run: |
    dd if=/dev/zero of=1mb.txt bs=1M count=1
    tar cvzf 1mb.tar.gz 1mb.txt
    ls -lh 1mb.tar.gz
    gzip -l 1mb.tar.gz
```
I would need to delete or edit the "queues" key to use for myself, given that I don't want whatever is defined as the user "gpu_queue." Instead, the recipe file should be agnostic to the executor:

```
version: 0.0.1

zlib_compression:
  type: script
  run: |
    dd if=/dev/zero of=1mb.txt bs=1M count=1
    tar cvzf 1mb.tar.gz 1mb.txt
    ls -lh 1mb.tar.gz
    gzip -l 1mb.tar.gz
```
But then specified to run under some environment by the user, or by the default configuration file enforced for the user. For example:

```bash
buildtest build -c zlib.yml -e gpu_local
```
This makes the recipes portable, which is essential if you want people to share.

Reply:
I foresee it would be difficult to have fully portable tests when it comes to writing elaborate tests. One use-case in HPC is around having buildtest to automate acceptance testing for each site. Certainly this requires site-specific test which can't be shared It was discussed in SC19 https://olcf.github.io/hpc-system-test-wg/events/sc19bof.html . It would make sense to have a repository that be used by sites to share tests they run for acceptance test for their sites. 

It gives HPC a common framework to write test, even if tests are site specific but allows anyone to understand the tests and adapt them to their site.

Some basic acceptance test could be simple, but not entirely portable 
- Check login nodes are reachable
- Check slurmctl and slurmdb
- Check filesystem mounts
- Run a series of dummy slurm jobs for sanity (batch, job-array, multi-nodes, gpu jobs)

Reply:
Yes I agree, but hard coding a test specification (executor) in a test recipe is a different thing. If anything needs to be hard coded, it should be in the site-specific (customized) configuration and the test would still be agnostic to that.